instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinInstrumentationModule.java

```diff
@@ -0,0 +1,330 @@
+import com.vaadin.flow.server.RequestHandler;
+import com.vaadin.flow.server.VaadinService;
+import com.vaadin.flow.server.communication.rpc.RpcInvocationHandler;
+import elemental.json.JsonObject;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.javaagent.tooling.InstrumentationModule;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+@AutoService(InstrumentationModule.class)
+public class VaadinInstrumentationModule extends InstrumentationModule {
+
+  public VaadinInstrumentationModule() {
+    super("vaadin", "vaadin-14.2");
+  }
+
+  @Override
+  public ElementMatcher.Junction<ClassLoader> classLoaderMatcher() {
+    // class added in vaadin 14.2
+    return hasClassesNamed("com.vaadin.flow.server.frontend.installer.NodeInstaller");
+  }
+
+  @Override
+  public List<TypeInstrumentation> typeInstrumentations() {
+    return asList(
+        new VaadinServiceInstrumentation(),
+        new RequestHandlerInstrumentation(),
+        new UiInstrumentation(),
+        new RouterInstrumentation(),
+        new JavaScriptBootstrapUiInstrumentation(),
+        new RpcInvocationHandlerInstrumentation(),
+        new ClientCallableRpcInstrumentation());
+  }
+
+  // add span around vaadin request processing code
+  public static class VaadinServiceInstrumentation implements TypeInstrumentation {
+
+    @Override
+    public ElementMatcher<? super TypeDescription> typeMatcher() {
+      return named("com.vaadin.flow.server.VaadinService");
+    }
+
+    @Override
+    public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+      return singletonMap(
+          named("handleRequest")
+              .and(takesArgument(0, named("com.vaadin.flow.server.VaadinRequest")))
+              .and(takesArgument(1, named("com.vaadin.flow.server.VaadinResponse"))),
+          HandleRequestAdvice.class.getName());
+    }
+
+    public static class HandleRequestAdvice {
+      @Advice.OnMethodEnter(suppress = Throwable.class)
+      public static void onEnter(
+          @Advice.This VaadinService vaadinService,
+          @Advice.Origin Method method,
+          @Advice.Local("otelContext") Context context,
+          @Advice.Local("otelScope") Scope scope) {
+        context = tracer().startVaadinServiceSpan(vaadinService, method);
+        scope = context.makeCurrent();
+      }
+
+      @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+      public static void onExit(
+          @Advice.Thrown Throwable throwable,
+          @Advice.Local("otelContext") Context context,
+          @Advice.Local("otelScope") Scope scope) {
+        if (scope == null) {
```

Comment:
`scope` cannot really be null here, can it?

Reply:
Fixed.